File: mycollab-services/src/main/java/com/esofthead/mycollab/module/crm/service/ibatis/LeadServiceImpl.java
/**
 * This file is part of mycollab-services.
 *
 * mycollab-services is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * mycollab-services is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mycollab-services.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.esofthead.mycollab.module.crm.service.ibatis;

import com.esofthead.mycollab.common.ModuleNameConstants;
import com.esofthead.mycollab.common.interceptor.aspect.ClassInfo;
import com.esofthead.mycollab.common.interceptor.aspect.ClassInfoMap;
import com.esofthead.mycollab.common.interceptor.aspect.Traceable;
import com.esofthead.mycollab.common.interceptor.aspect.Watchable;
import com.esofthead.mycollab.core.cache.CacheKey;
import com.esofthead.mycollab.core.persistence.ICrudGenericDAO;
import com.esofthead.mycollab.core.persistence.ISearchableDAO;
import com.esofthead.mycollab.core.persistence.service.DefaultService;
import com.esofthead.mycollab.module.crm.CrmTypeConstants;
import com.esofthead.mycollab.module.crm.dao.LeadMapper;
import com.esofthead.mycollab.module.crm.dao.LeadMapperExt;
import com.esofthead.mycollab.module.crm.domain.*;
import com.esofthead.mycollab.module.crm.domain.criteria.LeadSearchCriteria;
import com.esofthead.mycollab.module.crm.service.*;
import com.esofthead.mycollab.spring.ApplicationContextUtil;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Arrays;
import java.util.GregorianCalendar;

@Service
@Transactional
@Traceable(nameField = "lastname")
@Watchable(userFieldName = "assignuser")
public class LeadServiceImpl extends DefaultService<Integer, Lead, LeadSearchCriteria> implements LeadService {
    static {
        ClassInfoMap.put(LeadServiceImpl.class, new ClassInfo(ModuleNameConstants.CRM, CrmTypeConstants.LEAD));
    }

    private static final Logger LOG = LoggerFactory.getLogger(LeadServiceImpl.class);

    @Autowired
    private LeadMapper leadMapper;
    @Autowired
    private LeadMapperExt leadMapperExt;

    @Override
    public ICrudGenericDAO<Integer, Lead> getCrudMapper() {
        return leadMapper;
    }

    @Override
    public ISearchableDAO<LeadSearchCriteria> getSearchMapper() {
        return leadMapperExt;
    }

    @Override
    public SimpleLead findById(Integer leadId, Integer sAccountId) {
        return leadMapperExt.findById(leadId);
    }

    @Override
    public Integer saveWithSession(Lead lead, String username) {
        Integer result = super.saveWithSession(lead, username);
        if (lead.getExtraData() != null && (lead.getExtraData() instanceof SimpleCampaign)) {
            CampaignLead associateLead = new CampaignLead();
            associateLead.setCampaignid(((SimpleCampaign) lead.getExtraData()).getId());
            associateLead.setLeadid(lead.getId());
            associateLead.setCreatedtime(new GregorianCalendar().getTime());

            CampaignService campaignService = ApplicationContextUtil.getSpringBean(CampaignService.class);
            campaignService.saveCampaignLeadRelationship(Arrays.asList(associateLead), lead.getSaccountid());
        } else if (lead.getExtraData() != null && lead.getExtraData() instanceof SimpleOpportunity) {
            OpportunityLead associateLead = new OpportunityLead();
            associateLead.setOpportunityid(((SimpleOpportunity) lead.getExtraData()).getId());
            associateLead.setLeadid(lead.getId());
            associateLead.setCreatedtime(new GregorianCalendar().getTime());

            OpportunityService opportunityService = ApplicationContextUtil
                    .getSpringBean(OpportunityService.class);
            opportunityService.saveOpportunityLeadRelationship(
                    Arrays.asList(associateLead), lead.getSaccountid());
        }
        return result;
    }

    @Override
    public void convertLead(SimpleLead lead, Opportunity opportunity, String convertUser) {
        LOG.debug("Create new account and save it");
        Account account = new Account();
        account.setAccountname(lead.getAccountname());
        account.setNumemployees(lead.getNoemployees());
        account.setIndustry(lead.getIndustry());
        account.setEmail(lead.getEmail());
        account.setPhoneoffice(lead.getOfficephone());
        account.setFax(lead.getFax());
        account.setWebsite(lead.getWebsite());
        account.setAssignuser(lead.getAssignuser());
        account.setDescription(lead.getDescription());
        account.setSaccountid(lead.getSaccountid());

        AccountService accountService = ApplicationContextUtil.getSpringBean(AccountService.class);
        Integer accountId = accountService.saveWithSession(account, convertUser);

        LOG.debug("Create account lead relationship");
        AccountLead accLead = new AccountLead();
        accLead.setAccountid(accountId);
        accLead.setLeadid(lead.getId());
        accLead.setIsconvertrel(true);

        accountService.saveAccountLeadRelationship(Arrays.asList(accLead),
                lead.getSaccountid());

        LOG.debug("Create new contact and save it");
        Contact contact = new Contact();
        contact.setPrefix(lead.getPrefixname());
        contact.setFirstname(lead.getFirstname());
        contact.setLastname(lead.getLastname());
        contact.setTitle(lead.getTitle());
        contact.setDepartment(lead.getDepartment());
        contact.setAccountid(accountId);
        contact.setSaccountid(lead.getSaccountid());
        contact.setLeadsource(lead.getSource());
        contact.setAssignuser(lead.getAssignuser());
        contact.setOfficephone(lead.getOfficephone());
        contact.setEmail(lead.getEmail());
        contact.setSaccountid(lead.getSaccountid());

        ContactService contactService = ApplicationContextUtil
                .getSpringBean(ContactService.class);
        Integer contactId = contactService.saveWithSession(contact, convertUser);
        LOG.debug("Create contact lead relationship");
        ContactLead contactLead = new ContactLead();
        contactLead.setContactid(contactId);
        contactLead.setLeadid(lead.getId());
        contactLead.setIsconvertrel(true);
        contactService.saveContactLeadRelationship(Arrays.asList(contactLead),
                lead.getSaccountid());

        if (opportunity != null) {
            opportunity.setAccountid(accountId);
            opportunity.setSaccountid(lead.getSaccountid());
            OpportunityService opportunityService = ApplicationContextUtil.getSpringBean(OpportunityService.class);
            int opportunityId = opportunityService.saveWithSession(opportunity,
                    convertUser);

            LOG.debug("Create new opportunity contact relationship");
            ContactOpportunity oppContact = new ContactOpportunity();
            oppContact.setContactid(contactId);
            oppContact.setOpportunityid(opportunityId);
            contactService.saveContactOpportunityRelationship(
                    Arrays.asList(oppContact), lead.getSaccountid());

            LOG.debug("Create new opportunity lead relationship");
            OpportunityLead oppLead = new OpportunityLead();
            oppLead.setLeadid(lead.getId());
            oppLead.setOpportunityid(opportunityId);
            oppLead.setIsconvertrel(true);
            oppLead.setCreatedtime(new GregorianCalendar().getTime());
            opportunityService.saveOpportunityLeadRelationship(
                    Arrays.asList(oppLead), lead.getSaccountid());
        }

    }

    @Override
    public SimpleLead findConvertedLeadOfAccount(Integer accountId, @CacheKey Integer sAccountId) {
        return leadMapperExt.findConvertedLeadOfAccount(accountId);
    }

    @Override
    public SimpleLead findConvertedLeadOfContact(Integer contactId, @CacheKey Integer sAccountId) {
        return leadMapperExt.findConvertedLeadOfContact(contactId);
    }

    @Override
    public SimpleLead findConvertedLeadOfOpportunity(Integer opportunity, @CacheKey Integer sAccountId) {
        return leadMapperExt.findConvertedLeadOfOpportunity(opportunity);
    }
}

File: mycollab-services/src/main/java/com/mycollab/module/crm/service/impl/LeadServiceImpl.kt
package com.mycollab.module.crm.service.impl

import com.mycollab.aspect.ClassInfo
import com.mycollab.aspect.ClassInfoMap
import com.mycollab.aspect.Traceable
import com.mycollab.aspect.Watchable
import com.mycollab.common.ModuleNameConstants
import com.mycollab.core.cache.CacheKey
import com.mycollab.db.persistence.ICrudGenericDAO
import com.mycollab.db.persistence.ISearchableDAO
import com.mycollab.db.persistence.service.DefaultService
import com.mycollab.module.crm.CrmTypeConstants
import com.mycollab.module.crm.dao.LeadMapper
import com.mycollab.module.crm.dao.LeadMapperExt
import com.mycollab.module.crm.domain.*
import com.mycollab.module.crm.domain.criteria.LeadSearchCriteria
import com.mycollab.module.crm.service.*
import com.mycollab.spring.AppContextUtil
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
@Transactional
@Traceable(nameField = "lastname")
@Watchable(userFieldName = "assignuser")
class LeadServiceImpl(private val leadMapper: LeadMapper,
                      private val leadMapperExt: LeadMapperExt) : DefaultService<Int, Lead, LeadSearchCriteria>(), LeadService {

    override val crudMapper: ICrudGenericDAO<Int, Lead>
        get() = leadMapper as ICrudGenericDAO<Int, Lead>

    override val searchMapper: ISearchableDAO<LeadSearchCriteria>
        get() = leadMapperExt

    override fun findById(leadId: Int, sAccountId: Int): SimpleLead? {
        return leadMapperExt.findById(leadId)
    }

    override fun saveWithSession(lead: Lead, username: String?): Int {
        val result = super.saveWithSession(lead, username)
        if (lead.extraData != null && lead.extraData is SimpleCampaign) {
            val associateLead = CampaignLead()
            associateLead.campaignid = (lead.extraData as SimpleCampaign).id
            associateLead.leadid = lead.id
            associateLead.createdtime = GregorianCalendar().time

            val campaignService = AppContextUtil.getSpringBean(CampaignService::class.java)
            campaignService.saveCampaignLeadRelationship(listOf(associateLead), lead.saccountid)
        } else if (lead.extraData != null && lead.extraData is SimpleOpportunity) {
            val associateLead = OpportunityLead()
            associateLead.opportunityid = (lead.extraData as SimpleOpportunity).id
            associateLead.leadid = lead.id
            associateLead.createdtime = GregorianCalendar().time

            val opportunityService = AppContextUtil.getSpringBean(OpportunityService::class.java)
            opportunityService.saveOpportunityLeadRelationship(listOf(associateLead), lead.saccountid)
        }
        return result
    }

    override fun convertLead(lead: SimpleLead, opportunity: Opportunity?, convertUser: String) {
        LOG.debug("Create new account and save it")
        val account = Account()
        account.accountname = lead.accountname
        account.numemployees = lead.noemployees
        account.industry = lead.industry
        account.email = lead.email
        account.phoneoffice = lead.officephone
        account.fax = lead.fax
        account.website = lead.website
        account.assignuser = lead.assignuser
        account.description = lead.description
        account.saccountid = lead.saccountid

        val accountService = AppContextUtil.getSpringBean(AccountService::class.java)
        val accountId = accountService.saveWithSession(account, convertUser)

        LOG.debug("Create account lead relationship")
        val accLead = AccountLead()
        accLead.accountid = accountId
        accLead.leadid = lead.id
        accLead.isconvertrel = true

        accountService.saveAccountLeadRelationship(listOf(accLead), lead.saccountid)

        LOG.debug("Create new contact and save it")
        val contact = Contact()
        contact.prefix = lead.prefixname
        contact.firstname = lead.firstname
        contact.lastname = lead.lastname
        contact.title = lead.title
        contact.department = lead.department
        contact.accountid = accountId
        contact.saccountid = lead.saccountid
        contact.leadsource = lead.source
        contact.assignuser = lead.assignuser
        contact.officephone = lead.officephone
        contact.email = lead.email
        contact.saccountid = lead.saccountid

        val contactService = AppContextUtil.getSpringBean(ContactService::class.java)
        val contactId = contactService.saveWithSession(contact, convertUser)
        LOG.debug("Create contact lead relationship")
        val contactLead = ContactLead()
        contactLead.contactid = contactId
        contactLead.leadid = lead.id
        contactLead.isconvertrel = true
        contactService.saveContactLeadRelationship(listOf(contactLead), lead.saccountid)

        if (opportunity != null) {
            opportunity.accountid = accountId
            opportunity.saccountid = lead.saccountid
            val opportunityService = AppContextUtil.getSpringBean(OpportunityService::class.java)
            val opportunityId = opportunityService.saveWithSession(opportunity, convertUser)

            LOG.debug("Create new opportunity contact relationship")
            val oppContact = ContactOpportunity()
            oppContact.contactid = contactId
            oppContact.opportunityid = opportunityId
            contactService.saveContactOpportunityRelationship(listOf(oppContact), lead.saccountid)

            LOG.debug("Create new opportunity lead relationship")
            val oppLead = OpportunityLead()
            oppLead.leadid = lead.id
            oppLead.opportunityid = opportunityId
            oppLead.isconvertrel = true
            oppLead.createdtime = GregorianCalendar().time
            opportunityService.saveOpportunityLeadRelationship(listOf(oppLead), lead.saccountid)
        }

    }

    override fun findConvertedLeadOfAccount(accountId: Int?, @CacheKey sAccountId: Int?): SimpleLead {
        return leadMapperExt.findConvertedLeadOfAccount(accountId)
    }

    override fun findConvertedLeadOfContact(contactId: Int?, @CacheKey sAccountId: Int?): SimpleLead {
        return leadMapperExt.findConvertedLeadOfContact(contactId)
    }

    override fun findConvertedLeadOfOpportunity(opportunity: Int?, @CacheKey sAccountId: Int?): SimpleLead {
        return leadMapperExt.findConvertedLeadOfOpportunity(opportunity)
    }

    companion object {
        init {
            ClassInfoMap.put(LeadServiceImpl::class.java, ClassInfo(ModuleNameConstants.CRM, CrmTypeConstants.LEAD))
        }

        private val LOG = LoggerFactory.getLogger(LeadServiceImpl::class.java)
    }
}
